cpp/src/arrow/ipc/reader.h

```diff
@@ -190,6 +193,85 @@ class ARROW_EXPORT RecordBatchFileReader {
   virtual Status ReadRecordBatch(int i, std::shared_ptr<RecordBatch>* batch) = 0;
 };
 
+/// \class RecordBatchReceiverCollect
+/// \brief Collect record batches read by RecordBatchStreamEmitter.
+class ARROW_EXPORT RecordBatchReceiverCollect : public Receiver {
+ public:
+  RecordBatchReceiverCollect() : record_batches_() {}
+  virtual ~RecordBatchReceiverCollect() = default;
+
+  Status RecordBatchReceived(std::shared_ptr<RecordBatch> record_batch) override {
+    record_batches_.push_back(std::move(record_batch));
+    return Status::OK();
+  }
+
+  /// \return the all read record batches
+  std::vector<std::shared_ptr<RecordBatch>> record_batches() const {
+    return record_batches_;
+  }
+
+ private:
+  std::vector<std::shared_ptr<RecordBatch>> record_batches_;
+};
+
+/// \class RecordBatchStreamEmitter
+/// \brief Push style record batch stream reader that receives data
+/// from user.
+///
+/// This class reads the schema (plus any dictionaries) as the first
+/// messages in the fed data, followed by record batches.
+class ARROW_EXPORT RecordBatchStreamEmitter {
```

Comment:
I would call this `StreamDecoder` or something. At some point we'll add other methods to `Receiver`, so it won't emit just record batches.

Reply:
Right, we would add `TensorReceived` (or `OnTensor` or whatever the chosing naming is :-)).